alot/db/utils.py

```diff
@@ -107,122 +212,35 @@ def message_from_file(handle):
-                # the combined method is used, currently this prevents
-                # the interpretation of the recovered plain text
-                # mail. maybe that's a feature.
-                malformed = unicode(e)
-            else:
-                # parse decrypted message
-                n = message_from_string(d)
-
-                # add the decrypted message to m. note that n contains
-                # all the attachments, no need to walk over n here.
-                m.attach(n)
-
-                # add any defects found
-                m.defects.extend(n.defects)
-
-                # there are two methods for both signed and encrypted
-                # data, one is called 'RFC 1847 Encapsulation' by
-                # RFC 3156, and one is the 'Combined method'.
-                if len(sigs) == 0:
-                    # 'RFC 1847 Encapsulation', the signature is a
-                    # detached signature found in the recovered mime
-                    # message of type multipart/signed.
-                    if X_SIGNATURE_VALID_HEADER in n:
-                        for k in (X_SIGNATURE_VALID_HEADER,
-                                  X_SIGNATURE_MESSAGE_HEADER):
-                            m[k] = n[k]
-                    else:
-                        # an encrypted message without signatures
-                        # should arouse some suspicion, better warn
-                        # the user
-                        add_signature_headers(m, [], 'no signature found')
-                else:
-                    # 'Combined method', the signatures are returned
-                    # by the decrypt_verify function.
-
-                    # note that if we reached this point, we know the
-                    # signatures are valid. if they were not valid,
-                    # the else block of the current try would not have
-                    # been executed
-                    add_signature_headers(m, sigs, '')
-
-        if malformed:
-            msg = u'Malformed OpenPGP message: {0}'.format(malformed)
-            content = email.message_from_string(msg.encode('utf-8'))
-            content.set_charset('utf-8')
-            m.attach(content)
+    if m.is_multipart():
+        p = get_params(m)
+
+        # handle OpenPGP signed data
+        if (m.get_content_subtype() == 'signed' and
+                p.get('protocol') == _APP_PGP_SIG):
+            _handle_signatures(m, m, p)
+
+        # handle OpenPGP encrypted data
+        elif (m.get_content_subtype() == 'encrypted' and
+              p.get('protocol') == _APP_PGP_ENC and
+              'Version: 1' in m.get_payload(0).get_payload()):
+            _handle_encrypted(m, m)
+
+        # It is also possible to put either of the abov into a multipart/mixed
+        # segment
+        elif m.get_content_subtype() == 'mixed':
+            sub = m.get_payload(0)
+            p = get_params(sub)
+
+            if (sub.get_content_subtype() == 'signed' and
```

Comment:
This checks for content type `*/signed` but maybe it should check for `multipart/signed`? Same below for encrypted.